c-api/compat-5.3.c

```diff
@@ -336,23 +395,177 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+    lua_pushfstring(L, "attempt to load a %s chunk when 'mode' is '%s'", modename, mode);
+    return err;
+  }
+  return LUA_OK;
+}
+
+
+typedef struct compat53_LoadF {
+  int n;  /* number of pre-read characters */
+  FILE *f;  /* file being read */
+  char buff[COMPAT53_LUA_FILE_BUFFER_SIZE];  /* area for reading file */
+} compat53_LoadF;
+
+
+static const char *getF (lua_State *L, void *ud, size_t *size) {
+  compat53_LoadF *lf = (compat53_LoadF *)ud;
+  (void)L;  /* not used */
+  if (lf->n > 0) {  /* are there pre-read characters to be read? */
+    *size = lf->n;  /* return them (chars already in buffer) */
+    lf->n = 0;  /* no more pre-read characters */
+  }
+  else {  /* read a block from file */
+    /* 'fread' can return > 0 *and* set the EOF flag. If next call to
+       'getF' called 'fread', it might still wait for user input.
+       The next check avoids this problem. */
+    if (feof(lf->f)) return NULL;
+    *size = fread(lf->buff, 1, sizeof(lf->buff), lf->f);  /* read block */
+  }
+  return lf->buff;
+}
+
+
+static int compat53_errfile (lua_State *L, const char *what, int fnameindex) {
+  char buf[512] = {0};
+  const char *serr = compat53_strerror(errno, buf, sizeof(buf));
+  const char *filename = lua_tostring(L, fnameindex) + 1;
+  lua_pushfstring(L, "cannot %s %s: %s", what, filename, serr);
+  lua_remove(L, fnameindex);
+  return LUA_ERRFILE;
+}
+
+
+static int compat53_skipBOM (compat53_LoadF *lf) {
+  const char *p = "\xEF\xBB\xBF";  /* UTF-8 BOM mark */
+  int c;
+  lf->n = 0;
+  do {
+    c = getc(lf->f);
+    if (c == EOF || c != *(const unsigned char *)p++) return c;
+    lf->buff[lf->n++] = c;  /* to be read by the parser */
+  } while (*p != '\0');
+  lf->n = 0;  /* prefix matched; discard it */
+  return getc(lf->f);  /* return next character */
+}
+
+
+/*
+** reads the first character of file 'f' and skips an optional BOM mark
+** in its beginning plus its first line if it starts with '#'. Returns
+** true if it skipped the first line.  In any case, '*cp' has the
+** first "valid" character of the file (after the optional BOM and
+** a first-line comment).
+*/
+static int compat53_skipcomment (compat53_LoadF *lf, int *cp) {
+  int c = *cp = compat53_skipBOM(lf);
+  if (c == '#') {  /* first line is a comment (Unix exec. file)? */
+    do {  /* skip first line */
+      c = getc(lf->f);
+    } while (c != EOF && c != '\n');
+    *cp = getc(lf->f);  /* skip end-of-line, if present */
+    return 1;  /* there was a comment */
+  }
+  else return 0;  /* no comment */
+}
+
+
+COMPAT53_API int luaL_loadfilex (lua_State *L, const char *filename, const char *mode) {
+  static const char lua_signature[] = "\x1bLua";
+  compat53_LoadF lf;
+  int status, readstatus;
+  int c;
+  int fnameindex = lua_gettop(L) + 1;  /* index of filename on the stack */
+  if (filename == NULL) {
+    lua_pushliteral(L, "=stdin");
+    lf.f = stdin;
+  }
+  else {
+    lua_pushfstring(L, "@%s", filename);
+#if defined(_MSC_VER)
```

Comment:
Should this be conditional on` _MSC_VER`? or `_WIN32`?

`fopen_s` can have different behaviour than `fopen` (e.g. it holds a lock on the file); so I'm not sure we should be using it...

Reply:
_WIN32 would apply on MinGW, where the deprecation error doesn't exist: the compiler and its associated library is what's providing the error, not Windows.

Reply:
As for the locking, at that point I'd have to drop down to an even lower API, like `_fsopen` or similar. I don't have any other choice for VC++ users: the code will explicitly error them if they keep using `fopen` as-is, and if I they value the deprecation errors then it will be my library that gets tossed out the window for not playing ball.

Reply:
> As for the locking, at that point I'd have to drop down to an even lower API, like _fsopen or similar.

That might be better actually, looks like `_fsopen` with `_SH_DENYNO` is the same as `fopen`, but without the warning.

Reply:
Also, please add a comment here that the workaround is to *avoid a warning* in MSVC version XYZ. (rather than, e.g. working around an bug or platform idiosyncrasy)

Reply:
It's not just a warning: it's a hard error and a build won't pass in the default VC++ with it, on any warning level.

I added the checks and also default to using `_fsopen` here, but someone can forcefully opt out of it if they expect `fopen_s`'s behavior by default.

Reply:
> if they expect fopen_s's behavior by default.

Why would they expect this, when lua itself doesn't have this. => we're trying to give lua semantics here, not invent new ones (even if they're better).

Reply:
That's why it's an explicit opt-in with a macro: if someone wants this behavior, they would need to explicitly ask for it by definition of `COMPAT53_NO_FOPEN_LOCK` to `0`, which will prompt the typical `fopen_s` on their system. Otherwise, it behaves exactly like Lua / like plain `fopen` does.